File: valid-parentheses/Solution20.java
// https://leetcode.com/problems/valid-parentheses/

import java.util.*;

class Solution20 {

    private static final Set<Character> OPEN_SYMS = Set.of(
            '{', '[', '(');
    private static final Map<Character, Character> CLOSE_SYMS = Map.of(
            '}', '{',
            ']', '[',
            ')', '(');

    // T: O(n), S: O(n)
    public boolean isValid(String s) {
        var stack = new Stack<Character>();
        for (int i = 0; i < s.length(); i++) {
            char ch = s.charAt(i);
            if (OPEN_SYMS.contains(ch)) {
                stack.push(ch);
            } else if (CLOSE_SYMS.containsKey(ch)) {
                if (stack.isEmpty()) {
                    return false;
                }
                char top = stack.pop();
                if (CLOSE_SYMS.get(ch) != top) {
                    return false;
                }
            }
        }
        return stack.size() == 0;
    }

    public static void main(String[] args) {
        var sol = new Solution20();
        var s1 = "()";
        var s2 = "()[]{}";
        var s3 = "(]";
        System.out.println(sol.isValid(s1)); // true
        System.out.println(sol.isValid(s2)); // true
        System.out.println(sol.isValid(s3)); // false
    }

}

File: valid-parentheses/solution20.py
# https://leetcode.com/problems/valid-parentheses/

class Solution20:

    OPEN_SYMS = {'{', '[', '('}
    CLOSE_SYMS = {'}': '{', ']': '[', ')': '('}

    # T: O(n), S: O(n)
    def isValid(self, s: str) -> bool:
        stack = []
        for ch in s:
            if ch in self.OPEN_SYMS:
                stack.append(ch)
            elif ch in self.CLOSE_SYMS:
                if not stack or stack.pop() != self.CLOSE_SYMS[ch]:
                    return False
        return len(stack) == 0


def main() -> None:
    sol = Solution20()
    s1 = "()"
    s2 = "()[]{}"
    s3 = "(]"
    print(sol.isValid(s1))  # True
    print(sol.isValid(s2))  # True
    print(sol.isValid(s3))  # False


if __name__ == "__main__":
    main()
